### PR TITLE
[python-package] Add support for passing Arrow to LightGBM

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This PR adds Arrow-support to the Python API of LightGBM and, thus, (partially) fixes https://github.com/microsoft/LightGBM/issues/3369.

# Changes

- Allow to pass Arrow table to `lgb.Dataset.data` and `booster.predict`
- Allow to pass Arrow arrays to `lgb.Dataset`'s `label`, `group`, `weight`, and `init_score`
- Add tests for C++ and Python

### CODE REVIEW

**Comment by borchero** [comment]
@microsoft-github-policy-service agree company=QuantCo

**Comment by borchero** [comment]
@microsoft-github-policy-service agree company="QuantCo"

File: python-package/lightgbm/arrow.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Iterator, Union
+
+import pyarrow as pa
+from pyarrow.cffi import ffi
+
+
+@dataclass
+class ArrowCArray:
```
Comment (jameslamb) [inline python-package/lightgbm/arrow.py:12]:
```suggestion
class _ArrowCArray:
```

Since this is intended to be an internal implementation detail, please prefix it with a `_`.

File: python-package/lightgbm/arrow.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        return int(ffi.cast("uintptr_t", ffi.addressof(self.chunks[0])))
+
+    @property
+    def schema_ptr(self) -> int:
+        """Returns the address of the pointer to the schema of the array."""
+        return int(ffi.cast("uintptr_t", self.schema))
+
+
+@contextmanager
+def export_arrow_to_c(data: Union[pa.Table, pa.Array, pa.ChunkedArray]) -> Iterator[ArrowCArray]:
```
Comment (jameslamb) [inline python-package/lightgbm/arrow.py:31]:
```suggestion
def _export_arrow_to_c(data: Union[pa.Table, pa.Array, pa.ChunkedArray]) -> Iterator[ArrowCArray]:
```

File: python-package/lightgbm/arrow.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# coding: utf-8
+"""Utilities for handling Arrow in LightGBM."""
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Iterator, Union
+
+import pyarrow as pa
+from pyarrow.cffi import ffi
+
```
Comment (jameslamb) [inline python-package/lightgbm/arrow.py:9]:
```suggestion

__all__: List[str] = []
```

Please make it clear that nothing iin this module is intended to be part of the public API, e.g. that:

* it won't have dedicated entries in the docs generated with `sphinx` autodoc
* it won't be pulled in by `from lightgbm import *`

File: python-package/lightgbm/arrow.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+@dataclass
```
Comment (jameslamb) [inline python-package/lightgbm/arrow.py:11]: This project still support Python 3.6, so you cannot rely on `dataclasses` from the standard library being available: https://github.com/microsoft/LightGBM/pull/5765#issuecomment-1454861675

For the purpose of this PR, please just make it a normal class and add the bit of `__init__()` boilerplate like

```python
class _ArrowCArray:
    def __init__(self, n_chunks, chunks, schema):
        self.n_chunks = n_chunks
        self.chunks = chunks
        self.schema = schema
```

File: python-package/lightgbm/basic.py [hunk 5]
```diff
@@ -2394,15 +2494,15 @@ def _reverse_update_params(self) -> "Dataset":
...
+        data: Optional[Union[List[List[float]], List[List[int]], List[float], List[int], np.ndarray, pd_Series, pd_DataFrame, pa_Table, pa_Array, pa_ChunkedArray]]
     ) -> "Dataset":
         """Set property into the Dataset.
 
         Parameters
         ----------
         field_name : str
             The field name of the information.
-        data : list, list of lists (for multi-class task), numpy array, pandas Series, pandas DataFrame (for multi-class task), or None
+        data : list, list of lists (for multi-class task), numpy array, pandas Series, pandas DataFrame (for multi-class task), pyarrow Table, pyarrow Array, or None
```
Comment (jameslamb) [inline python-package/lightgbm/basic.py:2505]:
```suggestion
        data : list, list of lists (for multi-class task), numpy array, pandas Series, pandas DataFrame (for multi-class task), pyarrow Table, pyarrow Array, pyarrow ChunkedArray or None
```

Based on the corresponding change to the type hint, this should also note that a `ChunkedArray` is possible.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for working on this! I just gave this a very quick review and left a few small notes. Will try to give it a more thorough review in the coming days.

Please let us know if you need help with the failing CI jobs. I can say with confidence that most of them are related to the state of this PR, not other flakiiness in those tests.

One other question.... would you consider limiting the scope of this PR to just accepting `Arrow` types for the training data, and defer `init_score`, `weight`, and being able to predict on Arrow data to follow-up PRs? That'd reduce the scope of this a bit, which should reduce the effort to for us to provide a thoughtful review. One way I've seen that work well in the past is to keep a PR like this one with all of the changes up as a draft, to show the end state you want to get to, and submit individual smaller PRs with more focused changesets.

**Comment by lorentzenchr** [comment]
Arrow support would be really nice.
Just an idea: How about using [nanoarrow](https://arrow.apache.org/nanoarrow/0.2.0/index.html). Still early, but is is exactly meant for cases like here.

File: python-package/lightgbm/arrow.py (hunk 4 above)
Comment (jameslamb) [reply python-package/lightgbm/arrow.py:11]: Hey sorry @borchero ... now that #6048 has been merged, in this and the other PRs you can use `dataclasses` freely! We decided to take a Python-3.6-only dependency on the `dataclasses` backport.

**Comment by sheldonrong** [comment]
@borchero any plans to move this forward? thanks.

**Comment by jameslamb** [comment]
> any plans to move this forward

Development is happening in #6034 right now.

**Comment by borchero** [comment]
Logic of this PR is now fully implemented via #6034, #6163, #6164, #6166, #6167, #6168.

**Comment by jameslamb** [comment]
Now that there are just 2 PRs left in the initial work for this (#6168, #6210), I think this draft PR can be safely closed.

Thanks so much for all your help and patience @borchero , and for splitting this up into smaller and easier-to-review pieces,
